src/peergos/server/net/DHTHandler.java

```diff
@@ -1,120 +1,174 @@
+import peergos.shared.ipfs.api.*;
 import peergos.shared.storage.ContentAddressedStorage;
-import peergos.shared.util.*;
 import com.sun.net.httpserver.*;
+import peergos.shared.util.*;
 
 import java.io.*;
 import java.util.*;
 import java.util.function.*;
+import java.util.stream.*;
 
 public class DHTHandler implements HttpHandler
 {
     private final ContentAddressedStorage dht;
+    private final String apiPrefix;
 
-    public DHTHandler(ContentAddressedStorage dht) throws IOException
+    public DHTHandler(ContentAddressedStorage dht, String apiPrefix) throws IOException
     {
         this.dht = dht;
+        this.apiPrefix = apiPrefix;
+    }
+
+    public DHTHandler(ContentAddressedStorage dht) throws IOException {
+        this(dht, "/api/v0/");
+    }
+
+    private Map<String, List<String>> parseQuery(String query) {
+        if (query.startsWith("?"))
+            query = query.substring(1);
+        String[] parts = query.split("&");
+        Map<String, List<String>> res = new HashMap<>();
+        for (String part: parts) {
+            int sep = part.indexOf("=");
+            String key = part.substring(0, sep);
+            String value = part.substring(sep + 1);
+            res.putIfAbsent(key, new ArrayList<>());
+            res.get(key).add(value);
+        }
+        return res;
     }
 
     @Override
     public void handle(HttpExchange httpExchange) throws IOException {
         try {
             String path = httpExchange.getRequestURI().getPath();
-            if (path.startsWith("/dht/get")) {
-                String ipfsPrefix = "/dht/get/ipfs/";
-                if (!path.startsWith(ipfsPrefix))
-                    httpExchange.sendResponseHeaders(404, 0);
-                else {
-                    Multihash key = Multihash.fromBase58(path.substring(ipfsPrefix.length()));
-                    byte[] value = dht.get(key);
-                    new GetSuccess(key, httpExchange).accept(value);
-                }
-            } else {
-                InputStream in = httpExchange.getRequestBody();
-                DataInputStream din = new DataInputStream(in);
-                int type = din.readInt();
-                if (type == 0) {
-                    // PUT
-                    byte[] value = Serialize.deserializeByteArray(din, ContentAddressedStorage.MAX_OBJECT_LENGTH);
-                    byte[] owner = Serialize.deserializeByteArray(din, UserPublicKey.MAX_SIZE);
-                    // TODO check we care about owner
-                    List<Multihash> hashes = new ArrayList<>();
-                    int nlinks = din.readInt();
-                    for (int i = 0; i < nlinks; i++)
-                        hashes.add(new Multihash(Serialize.deserializeByteArray(din, 1024)));
+            if (! path.startsWith(apiPrefix))
+                throw new IllegalStateException("Unsupported api version, required: " + apiPrefix);
+            path = path.substring(apiPrefix.length());
```

Comment:
I looked into this, and because of the Java api requiring a HttpExchange this wouldn't actually encapsulate anything, The httpExchange contains the full path anyway. In essence to do this would require writing our own web framework, so basically jet all over again. So I think it's simpler how it is.